cmd/ndm_daemonset/probe/addhandler.go

```diff
@@ -216,3 +241,154 @@ func (pe *ProbeEvent) createBlockDeviceResourceIfNoHolders(bd blockdevice.BlockD
 }
+
+func (pe *ProbeEvent) upgradeBD(bd blockdevice.BlockDevice, bdAPIList *apis.BlockDeviceList) (bool, error) {
+	// if the device is a partition and parent is in use, then the event
+	// is skipped and no further processing is required.
+	if bd.DeviceAttributes.DeviceType == libudevwrapper.UDEV_PARTITION {
+		parentBD := pe.Controller.BDHierarchy[bd.DevPath]
+		if parentBD.DevUse.InUse {
+			klog.V(4).Infof("parent device: %s of %s is in use, hence ignoring event", parentBD.DevPath, bd.DevPath)
+			return false, nil
+		}
+	}
+
+	// if not in use, there is no need of upgrade.
+	if !bd.DevUse.InUse {
+		return true, nil
+	}
+
+	// device is in use by cstor or localpv
+	klog.V(4).Infof("device: %s in use by cstor / localPV. hence ignoring event", bd.DevPath)
+
+	// try to generate old UUID for the device
+	legacyUUID, isVirt := generateLegacyUUID(bd)
+	klog.V(4).Infof("tried generating legacy UUID: %s for device: %s", legacyUUID, bd.DevPath)
+
+	legacyBDResource := pe.Controller.GetExistingBlockDeviceResource(bdAPIList, legacyUUID)
+
+	// blockdevice with old uuid exists and is in use. The internal annotation will be added.
+	if legacyBDResource != nil {
+		bd.UUID = legacyUUID
+		klog.V(4).Infof("blockdevice resource with legacy UUID exists, adding internal annotaion")
+		// we used old algorithm for this BD, update the details, with old annotation
+		deviceInfo := pe.Controller.NewDeviceInfoFromBlockDevice(&bd)
+		if legacyBDResource.Annotations == nil {
+			legacyBDResource.Annotations = make(map[string]string)
+		}
+		legacyBDResource.Annotations[internalUUIDAnnotation] = legacyUUIDScheme
+		err := pe.Controller.PushBlockDeviceResource(legacyBDResource, deviceInfo)
+		if err != nil {
+			klog.Errorf("adding %s:%s annotation on %s failed. Error: %v", internalUUIDAnnotation, legacyUUIDScheme, bd.UUID, err)
+			return false, err
+		}
+
+	}
+	// If the resource does not exist, it can happen in 2 cases:
+	// 1. This device do not use the old legacyUUID scheme,
+	// 2. The device uses old scheme, but is a virtual drive and hostname or path has changed
+	//
+	// There can also be cases of manually created blockdevices, but those should be excluded by the filter.
+	klog.V(4).Infof("device: %s may be virtual / device uses new uuid scheme", bd.DevPath)
+
+	// try generating new UUID and if it exists, uses new scheme
+	// return and further processing required
+	uuid, ok := generateUUID(bd)
+	klog.V(4).Infof("tried generating new UUID: %s, for device: %s", uuid, bd.DevPath)
+	if ok {
+		// check if bd with new uuid exists
+		// uses the new algorithm,
+		if pe.Controller.GetExistingBlockDeviceResource(bdAPIList, uuid) != nil {
+			klog.V(4).Infof("device: %s uses GPT based uuid algorithm", bd.DevPath)
+			return true, nil
+		} else {
+			// should never reach this case.
+			// only reaches if manually created blockdevices are present.
+			// this should have been filtered out in filter configs, no firther processing required.
+			return false, nil
```

Comment:
in case of localpv which is using old blockdevice, which is virtual disk also, generateUUID will return true because of the filesystem. And we will return from here, right?We should proceed further in this case and create the new block device?

Reply:
Has been modified to only add annotation. New device wont be created. If at the time of upgrade the system is in correct state, i.e the correct BD-BDC relationships are there and used BD is in claimed state. It will only update with FSUUID annotation.